### PR TITLE
Added a submissionID to status and use it to group pod of the same run

### PR DESCRIPTION
**Pull request by liyinan926**
@skonto

### CODE REVIEW

**Comment by skonto** [comment]
@liyinan926 will give it a try shortly.

File: pkg/controller/sparkapplication/spark_pod_eventhandler.go [hunk 1]
```diff
@@ -49,7 +55,7 @@ func (s *sparkPodEventHandler) onPodUpdated(old, updated interface{}) {
 	if updatedPod.ResourceVersion == oldPod.ResourceVersion {
 		return
 	}
-	glog.V(2).Infof("Pod %s updated in namespace %s.", updatedPod.GetObjectMeta().GetName(), updatedPod.GetObjectMeta().GetNamespace())
+	glog.V(2).Infof("Pod %s updated in namespace %s.", updatedPod.GetName(), updatedPod.GetNamespace())
```
Comment (skonto) [inline pkg/controller/sparkapplication/spark_pod_eventhandler.go]: In the past these return empty strings that is why we used the GetObjectMeta thing... let's keep an eye on this.

File: pkg/controller/sparkapplication/spark_pod_eventhandler.go [hunk 2]
```diff
@@ -68,13 +74,24 @@ func (s *sparkPodEventHandler) onPodDeleted(obj interface{}) {
...
-		glog.V(2).Infof("Enqueuing SparkApplication %s for app update processing.", appKey)
-		s.enqueueFunc(appKey)
+	appName, exists := getAppName(pod)
+	if !exists {
+		return
 	}
+
+	if submissionID, exists := pod.Labels[config.SubmissionIDLabel]; exists {
+		app, err := s.applicationLister.SparkApplications(pod.GetNamespace()).Get(appName)
+		if err != nil || app.Status.SubmissionID != submissionID {
```
Comment (skonto) [inline pkg/controller/sparkapplication/spark_pod_eventhandler.go]: Recently we run into an issue where pods stuck in terminating status (several people have reported this on K8s project), the question is whether the operator should be responsible for cleaning stuff in the background. For now we can ignore this and keep submitting.

Comment (liyinan926) [reply pkg/controller/sparkapplication/spark_pod_eventhandler.go]: I think essentially we want to wait for the driver pod to be completely gone. This is actually handled in `validateSparkResourceDeletion` under the `PendingRerun` state. If `validateSparkResourceDeletion` returns false, the state will remain `PendingRerun` and the next resync loop will check again.

Comment (skonto) [reply pkg/controller/sparkapplication/spark_pod_eventhandler.go]: That means `validateSparkResourceDeletion` will always return a stuck pod. Should we add a configurable policy to handle this, like retry a force delete after N time passes?

**Comment by skonto** [comment]
@liyinan926 I run into this:
```
kubectl describe sparkapplications -n spark
Name:         spark-pi
Namespace:    spark
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"sparkoperator.k8s.io/v1beta1","kind":"SparkApplication","metadata":{"annotations":{},"name":"spark-pi","namespace":"spark"}...
API Version:  sparkoperator.k8s.io/v1beta1
Kind:         SparkApplication
Metadata:
  Creation Timestamp:  2019-03-19T20:00:16Z
  Generation:          1
  Resource Version:    2594323
  Self Link:           /apis/sparkoperator.k8s.io/v1beta1/namespaces/spark/sparkapplications/spark-pi
  UID:                 9d7031fd-4a81-11e9-b674-0288a0a0c3ec
Spec:
  Arguments:
    10000
  Deps:
  Driver:
    Annotations:
      Fiorano . Lightbend . Io / Port:    9100
      Fiorano . Lightbend . Io / Scrape:  true
    Cores:                                1
    Labels:
      Version:        2.4.0
    Memory:           1g
    Service Account:  spark-sa
  Executor:
    Annotations:
      Fiorano . Lightbend . Io / Port:    9100
      Fiorano . Lightbend . Io / Scrape:  true
    Cores:                                2
    Instances:                            2
    Labels:
      Version:            2.4.0
    Memory:               1g
  Image:                  lightbend/spark:2.1.0-OpenShift-2.4.0-rh
  Image Pull Policy:      Always
  Main Application File:  local:///opt/spark/examples/jars/spark-examples_2.11-2.4.0.jar
  Main Class:             org.apache.spark.examples.SparkPi
  Mode:                   cluster
  Monitoring:
    Expose Driver Metrics:    true
    Expose Executor Metrics:  true
    Prometheus:
      Config File:       /etc/metrics/conf/prometheus.yaml
      Jmx Exporter Jar:  /prometheus/jmx_prometheus_javaagent-0.3.1.jar
      Port:              9100
  Restart Policy:
    Type:         Never
  Spark Version:  2.4.0
  Type:           Scala
Status:
  Application State:
    State:  SUBMITTED
  Driver Info:
    Web UI Port:                 32292
    Web UI Service Name:         spark-pi-ui-svc
  Execution Attempts:            1
  Last Submission Attempt Time:  2019-03-19T20:01:21Z
  Submission Attempts:           1
  Submission ID:                 spark-pi-1553025678548580441
  Termination Time:              <nil>
Events:
  Type     Reason                               Age                From            Message
  ----     ------                               ----               ----            -------
  Normal   SparkApplicationAdded                85s                spark-operator  SparkApplication spark-pi was added, enqueuing it for submission
  Normal   SparkExecutorPending                 68s (x2 over 68s)  spark-operator  Executor spark-pi-1553025617735-exec-1 is pending
  Normal   SparkExecutorPending                 67s (x2 over 67s)  spark-operator  Executor spark-pi-1553025617735-exec-2 is pending
  Normal   SparkExecutorRunning                 63s                spark-operator  Executor spark-pi-1553025617735-exec-1 is running
  Normal   SparkExecutorRunning                 62s                spark-operator  Executor spark-pi-1553025617735-exec-2 is running
  Normal   SparkDriverRunning                   46s (x2 over 76s)  spark-operator  Driver spark-pi-driver is running
  Normal   SparkExecutorPending                 42s (x2 over 43s)  spark-operator  Executor spark-pi-1553025648754-exec-1 is pending
  Normal   SparkExecutorPending                 42s (x2 over 42s)  spark-operator  Executor spark-pi-1553025648754-exec-2 is pending
  Normal   SparkExecutorRunning                 37s                spark-operator  Executor spark-pi-1553025648754-exec-2 is running
  Normal   SparkExecutorRunning                 37s                spark-operator  Executor spark-pi-1553025648754-exec-1 is running
  Normal   SparkApplicationSpecUpdateProcessed  22s (x3 over 54s)  spark-operator  Successfully processed spec update for SparkApplication spark-pi
  Warning  SparkApplicationPendingRerun         20s (x3 over 54s)  spark-operator  SparkApplication spark-pi is pending rerun
  Normal   SparkApplicationSubmitted            17s (x4 over 82s)  spark-operator  SparkApplication spark-pi was submitted successfully
```
while the app is running the status was never updated:
```
kubectl get pods -n spark

spark-pi-1553025682813-exec-1                                     1/1       Running       0          1m
spark-pi-1553025682813-exec-2                                     1/1       Running       0          1m

spark-pi-driver                                                   1/1       Running       0          1m

```

**Comment by skonto** [comment]
It seems that Invalidating state clears the driver thus `driver pod not found` msg printed and the final pod updates for the new running pods are never passing:
```
	if submissionID, exists := pod.Labels[config.SubmissionIDLabel]; exists {
		app, err := s.applicationLister.SparkApplications(pod.GetNamespace()).Get(appName)
		if err != nil || app.Status.SubmissionID != submissionID {
			return
		}
	}
```

File: pkg/controller/sparkapplication/spark_pod_eventhandler.go (hunk 2 above)
Comment (liyinan926) [reply pkg/controller/sparkapplication/spark_pod_eventhandler.go]: While, a pod has a `terminationGracePeriod` after which the pod will be sent a SIGKILL. So once deletion of the pod is requested, the pod should be gone after certain amount of time.

**Comment by liyinan926** [comment]
> It seems that Invalidating state clears the driver thus `driver pod not found` msg printed and the final pod updates for the new running pods are not triggering the app updates for the same reason.

The invalidation is not clearing driver info. That `driver pod not found` log is benign and just means the driver pod has not been created yet. This could happen if the sync is triggered after the application state gets set to `Submitted`, but the driver pod has not started up running yet.

**Comment by skonto** [comment]
Ok my mistake `PendingRerun` does:
```
 else if status.AppState.State == v1beta1.PendingRerunState {
		status.SparkApplicationID = ""
		status.DriverInfo = v1beta1.DriverInfo{}
		status.ExecutorState = nil
```

**Comment by liyinan926** [comment]
`currentDriverState` is a local variable in `getAndUpdateAppState`, and set based on what the pod listing result is. If the driver pod has started up running, it will show up in the list result, and as a result, `currentDriverState` will be set.

**Comment by skonto** [comment]
If you check the logs the pod has started, also `kubectl` shows that, but `driver not found for SparkApplication` is printed all the time meaning  `currentDriverState` is nil. It never changes and state remains as `SUBMITTED`.

**Comment by liyinan926** [comment]
That's weird. It means the pod listing result doesn't include the driver pod.

**Comment by skonto** [comment]
Might be a transient error.

**Comment by skonto** [comment]
I just hit another one..
I deleted my CRD and run:
```
kubectl  apply -f pi-metrics.yaml
sleep 10
kubectl  apply -f pi-metrics2.yaml
sleep 10
kubectl  apply -f pi-metrics3.yaml
sleep 10
kubectl  apply -f pi-metrics.yaml
```

```
  Application State:
    Error Message:  Driver Pod not found
    State:          FAILED
  Driver Info:
    Pod Name:                    spark-pi-driver
    Web UI Port:                 31660
    Web UI Service Name:         spark-pi-ui-svc
  Execution Attempts:            1
  Last Submission Attempt Time:  2019-03-19T23:56:18Z
  Spark Application Id:          spark-f3e575fefe4444939ca9bf66a69bb464
  Submission Attempts:           1
  Submission ID:                 spark-pi-1553039775702310039
  Termination Time:              2019-03-19T23:56:26Z
Events:
  Type     Reason                               Age                 From            Message
  ----     ------                               ----                ----            -------
  Normal   SparkApplicationAdded                2m1s                spark-operator  SparkApplication spark-pi was added, enqueuing it for submission
  Normal   SparkDriverRunning                   90s (x3 over 112s)  spark-operator  Driver spark-pi-driver is running
  Normal   SparkApplicationSpecUpdateProcessed  88s (x3 over 110s)  spark-operator  Successfully processed spec update for SparkApplication spark-pi
  Warning  SparkApplicationPendingRerun         88s (x3 over 110s)  spark-operator  SparkApplication spark-pi is pending rerun
  Normal   SparkApplicationSubmitted            85s (x4 over 118s)  spark-operator  SparkApplication spark-pi was submitted successfully
  Warning  SparkApplicationFailed               85s                 spark-operator  SparkApplication spark-pi failed: Driver Pod not found
```
log: https://gist.github.com/skonto/6fefa215b236cb83abd5cce68ace627a

It gets stuck to `FAILED`....

Pods are running fine at the end:

```
kubectl get pods -n spark
NAME                                                              READY     STATUS        RESTARTS   AGE

spark-operator-ns-fdp-sparkoperator-695455567f-jhghc              1/1       Running       0          21m
spark-pi-1553039787910-exec-1                                     1/1       Running       0          59s
spark-pi-1553039787910-exec-2                                     1/1       Running       0          59s
spark-pi-driver                                                   1/1       Running       0          1m
```

**Comment by liyinan926** [comment]
Again, it's weird. It means the application was in a state that should have an existing driver pod but for some reason the driver pod didn't show up on the pod listing result.

**Comment by skonto** [comment]
I believe this is a different case because after the app update there is no `driver not found for SparkApplication` printed all over the place. That means the previous `currentDriverState` is kept around and a failed pod update caused the change to the state to be `FAILING Driver Pod not found` and then to the final `FAILED Driver Pod not found`. This happens while there is a submission and the previous state before that is `SUBMITTED`. Most likely the submission didnt manage to update the status in time.
I think there are two consecutive updates here that move the state to the permanent FAILED one.
I think there is no pod at the moment because the new pod hasnt started yet.
```
I0319 23:56:29.454104      11 sparkapp_metrics.go:144] Exporting metrics for spark-pi; old status: { spark-pi-1553039775702310039 0001-01-01 00:00:00 +0000 UTC 0001-01-01 00:00:00 +0000 UTC {spark-pi-ui-svc 31660    spark-pi-driver} {PENDING_RERUN } map[] 0 0} new status: { spark-pi-1553039786574481132 2019-03-19 23:56:29.396916509 +0000 UTC m=+1186.542531986 0001-01-01 00:00:00 +0000 UTC {spark-pi-ui-svc 31007    } {SUBMITTED } map[] 1 1}
I0319 23:56:29.454153      11 controller.go:252] Ending processing key: "spark/spark-pi"
I0319 23:56:29.454190      11 controller.go:245] Starting processing key: "spark/spark-pi"
I0319 23:56:29.454232      11 controller.go:522] Trying to update SparkApplication spark/spark-pi, from: [{spark-f3e575fefe4444939ca9bf66a69bb464 spark-pi-1553039775702310039 2019-03-19 23:56:18 +0000 UTC 2019-03-19 23:56:26 +0000 UTC {spark-pi-ui-svc 31660    spark-pi-driver} {FAILING Driver Pod not found} map[] 1 1}] to [{spark-f3e575fefe4444939ca9bf66a69bb464 spark-pi-1553039775702310039 2019-03-19 23:56:18 +0000 UTC 2019-03-19 23:56:26 +0000 UTC {spark-pi-ui-svc 31660    spark-pi-driver} {FAILED Driver Pod not found} map[] 1 1}]
I0319 23:56:29.454690      11 event.go:221] Event(v1.ObjectReference{Kind:"SparkApplication", Namespace:"spark", Name:"spark-pi", UID:"87dd2218-4aa2-11e9-b674-0288a0a0c3ec", APIVersion:"sparkoperator.k8s.io/v1beta1", ResourceVersion:"2641608", FieldPath:""}): type: 'Warning' reason: 'SparkApplicationFailed' SparkApplication spark-pi failed: Driver Pod not found
I0319 23:56:29.455003      11 controller.go:205] SparkApplication spark/spark-pi was updated, enqueueing it
I0319 23:56:29.494486      11 controller.go:205] SparkApplication spark/spark-pi was updated, enqueueing it
```

**Comment by liyinan926** [comment]
The most interesting part of the logs is as follows. There was an update to the app that caused it's state to be changed to `INVALIDATING`, and then to `PENDING_RERUN` (the line starting with `I0319 23:56:26.556801`). However, in the interim, there was also an attempt to change the application state from `RUNNING` to `FAILING`. This latter state change is because the driver pod got deleted when the app was updated and invalidated. The second update caused this line https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/pkg/controller/sparkapplication/controller.go#L334 to be triggered. It is clearly that the reason for the second update to start with a `RUNNING` state is stale cache. The operator uses an informer and a lister for getting and listing `SparkApplication` objects. The informer/lister maintains a cache internally. When the status of an application gets updated on the API server, the informer receives the update and updates the cache asynchronously. Because there were several updates in a very short amount of time, there's a good chance that some update read a stale state of the app.

```
I0319 23:56:26.542436      11 controller.go:522] Trying to update SparkApplication spark/spark-pi, from: [{spark-f3e575fefe4444939ca9bf66a69bb464 spark-pi-1553039775702310039 2019-03-19 23:56:18 +0000 UTC 0001-01-01 00:00:00 +0000 UTC {spark-pi-ui-svc 31660    spark-pi-driver} {INVALIDATING } map[] 1 1}] to [{ spark-pi-1553039775702310039 0001-01-01 00:00:00 +0000 UTC 0001-01-01 00:00:00 +0000 UTC {spark-pi-ui-svc 31660    spark-pi-driver} {PENDING_RERUN } map[] 0 0}]
I0319 23:56:26.556801      11 sparkapp_metrics.go:144] Exporting metrics for spark-pi; old status: {spark-f3e575fefe4444939ca9bf66a69bb464 spark-pi-1553039775702310039 2019-03-19 23:56:18 +0000 UTC 0001-01-01 00:00:00 +0000 UTC {spark-pi-ui-svc 31660    spark-pi-driver} {INVALIDATING } map[] 1 1} new status: { spark-pi-1553039775702310039 0001-01-01 00:00:00 +0000 UTC 0001-01-01 00:00:00 +0000 UTC {spark-pi-ui-svc 31660    spark-pi-driver} {PENDING_RERUN } map[] 0 0}
I0319 23:56:26.556841      11 controller.go:252] Ending processing key: "spark/spark-pi"
I0319 23:56:26.556871      11 controller.go:245] Starting processing key: "spark/spark-pi"
W0319 23:56:26.556953      11 controller.go:334] driver not found for SparkApplication: spark/spark-pi
I0319 23:56:26.556975      11 controller.go:522] Trying to update SparkApplication spark/spark-pi, from: [{spark-f3e575fefe4444939ca9bf66a69bb464 spark-pi-1553039775702310039 2019-03-19 23:56:18 +0000 UTC 0001-01-01 00:00:00 +0000 UTC {spark-pi-ui-svc 31660    spark-pi-driver} {RUNNING } map[] 1 1}] to [{spark-f3e575fefe4444939ca9bf66a69bb464 spark-pi-1553039775702310039 2019-03-19 23:56:18 +0000 UTC 2019-03-19 23:56:26.556966857 +0000 UTC m=+1183.702582450 {spark-pi-ui-svc 31660    spark-pi-driver} {FAILING Driver Pod not found} map[] 1 1}]
I0319 23:56:26.558776      11 controller.go:205] SparkApplication spark/spark-pi was updated, enqueueing it
```

**Comment by skonto** [comment]
ok @liyinan926 let's merge this and take a note. Should we ping the k8s project for the limitation eg. open a bug?

**Comment by liyinan926** [comment]
@skonto this is a known limitation of the informer and the caching mechanism. There's always some amount of delay because of the async nature.

**Comment by skonto** [comment]
>  It is clearly that the reason for the second update to start with a RUNNING state is stale cache.
I am not sure about this because things run in parallel and the last part about SUBMITTED state is ambiguous.

**Comment by liyinan926** [comment]
> > It is clearly that the reason for the second update to start with a RUNNING state is stale cache.
> 
> I am not sure about this because things run in parallel and the last part about SUBMITTED state is ambiguous. If you get an update while on another thread the submission is happening is the code 100% safe (even if the cache was working fine)? Anyway we can always merge and re-open if something fails.

Actually update events are put into a queue and processed sequentially. However, because the cache is async in nature, a subsequent update may see a state that is stale. In this particularly case, the second update was trying to change the app state from `RUNNING` to `FAILED` as the log shows.

**Comment by skonto** [comment]
There are workers who dequeue items right?

**Comment by liyinan926** [comment]
> There are workers at the controller level what dequeue items, these run in parallel, no?
> processNextItem -> syncSparkApplication is called in parallel from what I see.

Oh, you are right. I forgot that there are multiple worker threads.

**Comment by skonto** [comment]
@liyinan926 are you going to work on this on another PR?

**Comment by liyinan926** [comment]
> @liyinan926 are you going to work on these corner cases on another PR?

By corner cases, do you mean handling reads from stale cache? If yes, the best possible solution is to always read from the API server. But that risks putting a much bigger pressure onto the API server, which is what the caching mechanism tries to alleviate.

**Comment by skonto** [comment]
No the cases where things fail because of the parallel execution.

**Pull request merged**